### PR TITLE
resolves #3519 compute highlight line ranges on source block relative to value of start attribute

### LINKED ISSUE #3519: highlight lines on source block should be relative to start value
If a start value is set, then the lines to highlight specified by the highlight attribute should be relative to the start value. In other words, if the start value is 5 and highlight=5, then the first line should be highlighted. Otherwise, it gets confusing which lines are being highlighted.

```
[source%linenums,ruby,start=5,highlight=6]
----
get {
  render "Hello, World!"
}
----
```

Note that this scenario is only relevant when linenums are enabled.

### PR DESCRIPTION
**Pull request by Mogztter**
resolves #3519

### CODE REVIEW

File: lib/asciidoctor/substitutors.rb [hunk 1]
```diff
@@ -971,11 +971,14 @@ def highlight_source source, process_callouts
   #
   # source - The String source.
   # spec   - The lines specifier (e.g., "1-5, !2, 10" or "1..5;!2;10")
+  # start  - The start attribute
   #
   # Returns an [Array] of unique, sorted line numbers.
-  def resolve_lines_to_highlight source, spec
+  def resolve_lines_to_highlight source, spec, start
```
Comment (mojavelinux) [inline lib/asciidoctor/substitutors.rb:977]: This method is used by Asciidoctor PDF, so this argument needs to be optional.

File: lib/asciidoctor/substitutors.rb [hunk 2]
```diff
@@ -948,7 +948,7 @@ def highlight_source source, process_callouts
     if (linenums_mode = (attr? 'linenums') ? (doc_attrs[%(#{syntax_hl_name}-linenums-mode)] || :table).to_sym : nil)
       start_line_number = 1 if (start_line_number = (attr 'start', 1).to_i) < 1
     end
-    highlight_lines = resolve_lines_to_highlight source, (attr 'highlight') if attr? 'highlight'
+    highlight_lines = resolve_lines_to_highlight source, (attr 'highlight'), start_line_number if attr? 'highlight'
```
Comment (mojavelinux) [inline lib/asciidoctor/substitutors.rb:951]: I'm trying to figure out if it makes more sense to pass in the start value, or to remove it from each value in the result. I suppose it's elegant to pass it in.

File: lib/asciidoctor/substitutors.rb [hunk 3]
```diff
@@ -971,11 +971,14 @@ def highlight_source source, process_callouts
...
   # spec   - The lines specifier (e.g., "1-5, !2, 10" or "1..5;!2;10")
+  # start  - The start attribute
   #
   # Returns an [Array] of unique, sorted line numbers.
-  def resolve_lines_to_highlight source, spec
+  def resolve_lines_to_highlight source, spec, start
     lines = []
     spec = spec.delete ' ' if spec.include? ' '
+    # If a start value is set, then the lines to highlight specified by the provided spec should be relative to the start value.
+    offset = start ? start - 1 : 0
```
Comment (mojavelinux) [inline lib/asciidoctor/substitutors.rb:981]: Let's keep the value nil if the start is 1. That why we avoid an unnecessary calculation per entry.

Comment (mojavelinux) [reply lib/asciidoctor/substitutors.rb:981]: I think it makes more sense to remove the offset from each entry after the lines have been resolved. That reduces it to a single place in the code.

```
start == 1 ? lines.sort : (lines.sort.map {|it| it - start + 1 })
```

File: lib/asciidoctor/substitutors.rb (hunk 1 above)
Comment (mojavelinux) [reply lib/asciidoctor/substitutors.rb:977]: `start = 1`

Comment (Mogztter) [reply lib/asciidoctor/substitutors.rb:977]: > This method is used by Asciidoctor PDF, so this argument needs to be optional.

The value is not optional by default? In other words, the value will be `nil` if not defined no?

> start = 1

Ok :+1:

Comment (mojavelinux) [reply lib/asciidoctor/substitutors.rb:977]: Without the default value, it is required (unlike JavaScript) and thus will throw a runtime error.

File: lib/asciidoctor/substitutors.rb (hunk 2 above)
Comment (Mogztter) [reply lib/asciidoctor/substitutors.rb:951]: `start_line_number` can be `nil`, should we handle this case in `resolve_lines_to_highlight`?

```rb
start && start > 1 ? (lines.sort.map {|it| it - start + 1 }) : lines.sort
```

As an alternative, we could state in the documentation that `start` must be a positive integer greater or equals to 1?

File: lib/asciidoctor/substitutors.rb (hunk 3 above)
Comment (Mogztter) [reply lib/asciidoctor/substitutors.rb:981]: That's a good idea.
I wonder if we should check that we won't produce negative numbers if start is greater than highlight?

Comment (mojavelinux) [reply lib/asciidoctor/substitutors.rb]: I don't think we should worry about negative numbers. They will effectively be ignored anyway by the syntax highlighter.

Comment (Mogztter) [reply lib/asciidoctor/substitutors.rb]: Indeed, I've added a test to make sure that the code is robust.

File: lib/asciidoctor/substitutors.rb (hunk 2 above)
Comment (mojavelinux) [reply lib/asciidoctor/substitutors.rb:949]: Good point. In that case, the method argument should be `start = nil` and the logic should be:

```ruby
unless (shift = start ? start - 1 : 0) == 0
  lines = lines.map {|it| it - shift }
end
lines.sort
```

**Comment by mojavelinux** [comment]
Nice work!

File: CHANGELOG.adoc [hunk 4]
```diff
@@ -37,6 +37,7 @@ Bug Fixes::
   * Escape ellipsis at start of line in manpage output (#3645) (*@jnavila*)
   * Don't register footnote with ID if a footnote is already registered with that ID (#3690)
   * Honor start attribute on ordered list in manpage output (#3714)
+  * Highlight lines relative to the start value (#3701) (*@mogztter*)
```
Comment (mojavelinux) [inline CHANGELOG.adoc:40]: Can you update this to read:

```
Compute highlight line ranges on source block relative to value of start attribute
```

Comment (Mogztter) [reply CHANGELOG.adoc]: Sure, done :smile:

Comment (mojavelinux) [reply CHANGELOG.adoc]: Thanks!

**Comment by mojavelinux** [comment]
Can you rebase one more time?

**Comment by Mogztter** [comment]
@mojavelinux Rebased on master and commits squashed.

**Pull request merged**